text/0000-router-service.md

```diff
@@ -0,0 +1,252 @@
+- Start Date: 2015-09-24
+- RFC PR:
+- Ember Issue:
+
+# Summary
+
+This RFC proposes:
+
+ - creating a public `router` service that is a superset of today's `Ember.Router`.
+ 
+ - codifying and expanding the supported public API for the `transition` object that is currently passed to `Route` hooks.
+
+These topics are closely related because they share a unified `RouteInfo` type, which will be described in detail.
+
+# Motivation
+
+Given the modern Ember concepts of Components and Services, it is clear that routing capability should be exposed as a Service. I hope this is uncontroversial, given that we already implement it as a service internally, and given that usage of these nominally-private APIs is already becoming widespread.
+
+The immediate benefit of having a `RouterService` is that you can inject it into components, giving them a friendly way to initiate transitions and ask questions about the current global router state.
+
+A second benefit is that we have the opportunity to add new capabilities to the `RouterService` to replace several common patterns in the wild that dive into private internals in order to get things done. There are several places where we leak internals from router.js, and we can plug those leaks.
+
+# Detailed design
+
+## RouterService
+
+By way of a simple example, the router service behaves like this:
+
+```js
+import Component from 'ember-component';
+import service from 'ember-service/inject';
+
+export default Component.extend({
+  router: service(),
+  actions: {
+    goToMars() {
+      this.get('router').transitionTo('planet.mars');
+    }
+  }
+});
+```
+
+Like any Service, it can also be injected into Helpers, Routes, etc.
+
+### Relationship between EmberRouter and RouterService
+
+Q: "Why are you calling this thing 'router' when we already have a router? Shouldn't the new thing be called 'routing' or something else?".
+
+A: We shouldn't have two things. From the user's perspective, there is just "the router", and it happens to be available as a service. While we're free to continue implementing it as multiple classes under the hood, the public API should present as a single, coherent concept.
+
+Terminology:
+
+ - `EmberRouter` is the class that we already have today, defined in `ember-routing/system/router` and available publicly as `Ember.Router`
+ - `RouterService` is the new class I am proposing.
+
+`EmberRouter` has the following public API today:
+
+ - `map`
+ - `location`
+ - `rootURL`
+ - `willTransition`
+ - `didTransition`
+
+That API will be carried over verbatim to `RouterService`, and the publicly accessible `Ember.Router` class will *become* `RouterService`. In terms of implementation, I expect the existing `EmberRouter` class will continue to exist mostly unchanged. But public access to it will be moderated through `RouterService`.
+
+### New Methods: Initiating Transitions
+
+```js
+transitionTo(routeName, ...models, queryParams)
+replaceWith(routeName, ...models, queryParms)
```

Comment:
I think it might be useful to also introduce more APIs here to get closer to parody with the underlying browser API. For instance it's currently not very clear how to go back a transition. Solving this in user space normally requires you to save off the previous transition to create an adhoc linked list across your app or create a service to keep track of this, which may be subject to getting out of sync. Things get more leaky when you are in a node environment and now need to guard against `window` references because the framework didn't provide a good way of handle cases.